tsdb/engine/tsm1/engine.go

```diff
@@ -700,111 +698,87 @@ func (e *Engine) Export(w io.Writer, basePath string, start time.Time, end time.
+				return err
+			}
 		}
-	}
 
-	for _, f := range whole {
-		if err := e.writeFileToBackup(f, basePath, filepath.Join(path, f), tw); err != nil {
-			return err
+		// if this TSM file had a tombstone we'll write out the whole thing too.
+		if tombstonePath != "" {
+			if err := e.writeFileToGzip(tombstonePath, basePath, filepath.Join(path, tombstonePath), zw); err != nil {
+				return err
+			}
 		}
+		// when writing a file to Gzip, zw.Close() is how you signify you are done writing the file.
+		// so the helper functions will do the closing.
+		// to write the next file to the gzip stream, we just reset the gzip writer.
+		// later, this zipped stream must be read with zr.Multistream(false) so that it will stop reading on each new file.
+		zw.Reset(w)
 	}
 
 	return nil
 }
 
-func (e *Engine) filterFileToBackup(name, shardRelativePath, fullPath string, start, end int64, tw *tar.Writer) error {
+func (e *Engine) filterFileToGzip(r *TSMReader, name, shardRelativePath, fullPath string, start, end int64, zw *gzip.Writer) error {
 	f, err := os.Stat(fullPath)
 	if err != nil {
 		return err
 	}
 
-	h := &tar.Header{
-		Name:    filepath.ToSlash(filepath.Join(shardRelativePath, name)),
-		ModTime: f.ModTime(),
-		Size:    f.Size(),
-		Mode:    int64(f.Mode()),
-	}
-	if err := tw.WriteHeader(h); err != nil {
-		return err
-	}
+	zw.Name = filepath.ToSlash(filepath.Join(shardRelativePath, name))
+	zw.ModTime = f.ModTime()
+
 	fr, err := os.Open(fullPath)
 	if err != nil {
 		return err
 	}
-
 	defer fr.Close()
 
-	// TODO (Adam):  instead of io.copy as in writeFileToBackup, we will need to:
-
-	// 2.  write the blocks we want to keep
-	// 3.  write an index?????
-	// 4.  write a footer
-
-	// will probably need most of this
-	r, err := NewTSMReader(fr)
-	if err != nil {
-		return err
-	}
-	defer r.Close()
-
-	w, err := NewTSMWriter(tw)
+	w, err := NewTSMWriter(zw)
 	if err != nil {
 		return err
 	}
-	defer func() {
-		closeErr := w.Close()
-		if err == nil {
-			err = closeErr
-		}
-
-		// Check for errors where we should not remove the file
-		// REVIEWER:  do I need to check anything here? lifted this code from compact.go
-
-		if err != nil {
-			w.Remove()
-		}
-	}()
+	defer w.Close()
 
 	// implicit else: here we iterate over the blocks and only keep the ones we really want.
 	var bi *BlockIterator
 	bi = r.BlockIterator()
 
 	for bi.Next() {
-		// not concerned with typ of checksum since we are just blindly writing back, with no decoding
+		// not concerned with typ or checksum since we are just blindly writing back, with no decoding
 		key, minTime, maxTime, _, _, buf, err := bi.Read()
 		if err != nil {
 			return err
 		}
 		if minTime >= start && minTime <= end ||
 			maxTime >= start && maxTime <= end {
-			w.WriteBlock(key, minTime, maxTime, buf)
+			err := w.WriteBlock(key, minTime, maxTime, buf)
+			if err != nil {
+				return err
+			}
 		}
 	}
-	w.WriteIndex()
+
```

Comment:
There is a recent change in #8981 which adds an `Err` func for `BlockIterator`.  You'll need to ensure that is `nil` before writing the index and returning success.

```
if err := bi.Err(); err != nil {
   return err
}
```